source/server/admin/stats_request.cc

```diff
@@ -76,6 +80,10 @@ bool StatsRequest::nextChunk(Buffer::Instance& response) {
   const uint64_t starting_response_length = response.length();
   while (response.length() - starting_response_length < chunk_size_) {
     while (stat_map_.empty()) {
+      if (params_.type_ != StatsType::All && phase_ == Phase::CountersAndGauges) {
+        renderPerHostMetrics(response);
```

Comment:
Wait, why are we calling renderPerHostMetrics() both here and on line 103? I think we just want to call that function exactly once per request, and line 103 looks like the right place, when we are exiting the CountersAndGauges phase.

I'm a little confused though, why this does not cause a test to fail. Maybe we are just checking the presence of per-host stats, but not checking for duplicates? Have you tried this interactively?

I'm not ruling out the possiblity here that I'm overlooking the reason we need the call here (which could happen an arbitrary number of times).

Reply:
This was a fix for a bug I found while writing the speed test code: when there is by-type filtering (eg show only Counters), the per-host metrics weren't emitted at all.

The cause is that 4 lines down from here is: 

```
      if (params_.type_ != StatsType::All) {
        render_->finalize(response);
        return false;
      }
```

I don't know if this is the right fix for that, but the block 4 lines down causes the call on line 103 to not get called. And this block needs to be before the next stanza of 

```
      if (phase_stat_count_ == 0) {
        render_->noStats(response, phase_string_);
      }
```

Is there a cleaner way to handle this case? If not, I'll at least add a comment here, because this is hard to understand without some explanation.

Reply:
I added a comment around there to explain what's happening; let me know if that is clear enough.

Or alternately, is there a way to refactor this so the call only needs to be in one place? I couldn't figure out a clean way to do that, but I'm open to ideas.

Reply:
possibly. But I'm wondering if my hypothesis is correct: that the way you have it you can issue per-endpoint stats multiple times for the same cluster.

Reply:
If that's the case then the first thing to do is write a test-case that repros my suspected duplicate and then we can iterate on the code to fix it.

Reply:
I don't think it can happen multiple times:

```
      if (params_.type_ != StatsType::All && phase_ == Phase::CountersAndGauges) {
        renderPerHostMetrics(response);
      }

      if (phase_stat_count_ == 0) {
        render_->noStats(response, phase_string_);
      } else {
        phase_stat_count_ = 0;
      }

      if (params_.type_ != StatsType::All) {
        render_->finalize(response);
        return false;
      }
```

In the case where `params_.type_ != StatsType::All` is true, first we'll do `renderPerHostMetrics()` if it's the right type, then we'll figure out `noStats()` or not, then we'll `return`. I don't see any way we can do `renderPerHostMetrics()` twice.

Reply:
Would it be clearer if it were refactored like this? 

```
    while (stat_map_.empty()) {
      if (params_.type_ != StatsType::All) {
        if (phase_ == Phase::CountersAndGauges) {
          // In the case of filtering by type, we need to call this before checking for
          // no stats in the phase, and then after that this function returns without the normal
          // advancing to the next phase.
          renderPerHostMetrics(response);
        }

        if (phase_stat_count_ == 0) {
          render_->noStats(response, phase_string_);
        }

        render_->finalize(response);
        return false;
      }

      if (phase_stat_count_ == 0) {
        render_->noStats(response, phase_string_);
      }

      phase_stat_count_ = 0;
      switch (phase_) {
```

Reply:
I think I like this option better, yes. Thanks!

Reply:
Ok, pushed that version